lang/funcs/collection_test.go

```diff
@@ -222,3 +222,556 @@ func TestLength(t *testing.T) {
+				return
+			} else if err != nil {
+				t.Fatalf("unexpected error: %s", err)
+			}
+
+			if !got.RawEquals(test.Want) {
+				t.Errorf("wrong result\ngot:  %#v\nwant: %#v", got, test.Want)
+			}
+		})
+	}
+}
+
+func TestCompact(t *testing.T) {
+	tests := []struct {
+		List cty.Value
+		Want cty.Value
+		Err  bool
+	}{
+		{
+			cty.ListVal([]cty.Value{
+				cty.StringVal("test"),
+				cty.StringVal(""),
+				cty.StringVal("test"),
+			}),
+			cty.ListVal([]cty.Value{
+				cty.StringVal("test"),
+				cty.StringVal("test"),
+			}),
+			false,
+		},
+		{
+			cty.ListVal([]cty.Value{
+				cty.StringVal(""),
+				cty.StringVal(""),
+				cty.StringVal(""),
+			}),
+			cty.ListValEmpty(cty.String),
+			false,
+		},
+		{
+			cty.ListValEmpty(cty.String),
+			cty.ListValEmpty(cty.String),
+			false,
+		},
+		{
+			cty.ListVal([]cty.Value{
+				cty.StringVal("test"),
+				cty.StringVal("test"),
+				cty.StringVal(""),
+			}),
+			cty.ListVal([]cty.Value{
+				cty.StringVal("test"),
+				cty.StringVal("test"),
+			}),
+			false,
+		},
+		{ // errors on list of lists
+			cty.ListVal([]cty.Value{
+				cty.ListVal([]cty.Value{
+					cty.StringVal("test"),
+				}),
+				cty.ListVal([]cty.Value{
+					cty.StringVal(""),
+				}),
+			}),
+			cty.NilVal,
+			true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("compact(%#v)", test.List), func(t *testing.T) {
+			got, err := Compact(test.List)
+
+			if test.Err {
+				if err == nil {
+					t.Fatal("succeeded; want error")
+				}
+				return
+			} else if err != nil {
+				t.Fatalf("unexpected error: %s", err)
+			}
+
+			if !got.RawEquals(test.Want) {
+				t.Errorf("wrong result\ngot:  %#v\nwant: %#v", got, test.Want)
+			}
+		})
+	}
+}
+
+func TestContains(t *testing.T) {
+	listOfStrings := cty.ListVal([]cty.Value{
+		cty.StringVal("the"),
+		cty.StringVal("quick"),
+		cty.StringVal("brown"),
+		cty.StringVal("fox"),
+	})
+	listOfInts := cty.ListVal([]cty.Value{
+		cty.NumberIntVal(1),
+		cty.NumberIntVal(2),
+		cty.NumberIntVal(3),
+		cty.NumberIntVal(4),
+	})
+
+	tests := []struct {
+		List  cty.Value
+		Value cty.Value
+		Want  cty.Value
+		Err   bool
+	}{
+		{
+			listOfStrings,
+			cty.StringVal("the"),
+			cty.BoolVal(true),
+			false,
+		},
+		{
+			listOfStrings,
+			cty.StringVal("penguin"),
+			cty.BoolVal(false),
+			false,
+		},
+		{
+			listOfInts,
+			cty.NumberIntVal(1),
+			cty.BoolVal(true),
+			false,
+		},
+		{
+			listOfInts,
+			cty.NumberIntVal(42),
+			cty.BoolVal(false),
+			false,
+		},
+		{ // And now we mix and match
+			listOfInts,
+			cty.StringVal("1"),
+			cty.BoolVal(false),
+			false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("contains(%#v, %#v)", test.List, test.Value), func(t *testing.T) {
+			got, err := Contains(test.List, test.Value)
+
+			if test.Err {
+				if err == nil {
+					t.Fatal("succeeded; want error")
+				}
+				return
+			} else if err != nil {
+				t.Fatalf("unexpected error: %s", err)
+			}
+
+			if got != test.Want {
```

Comment:
I guess this is working here because the two `cty.Value`s here are both primitives, but it isn't safe in general to use `==` and `!=` with these because they can contain values that Go can't compare directly, like slices. I'd suggest using `!got.RawEquals(test.Want)` in all of the tests just to avoid creating the impression for future maintainers that comparing values like this is possible in all cases.